### ISSUE TITLE
ESP8266WiFiClass refactoring

### ISSUE THREAD

**Issue by igrr** [esp8266/Arduino#850]
Logic for switching between different modes in ESP8266WiFiClass is spread across several functions and is hard to follow. Need to document how it *should* work, and then implement cleanly. This also applies to other processes which have state transitions — scanning, WPS, esptouch.

**Comment by tim-in-oakton** [comment]
I'm not sure that a refactoring is required, some well documented examples would likely make it workable without breaking any existing code.

**Comment by devyte** [comment]
@igrr is this refactoring still something that should be done?

**Comment by igrr** [comment]
Yes, once we have automated tests of some sort.
